### ISSUE TITLE
bulkInsert should return array

### ISSUE THREAD

**Issue by melnikaite** [sequelize/sequelize#14325]
## Issue Creation Checklist

- [x] I have read the [contribution guidelines](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)

## Bug Description

```js
module.exports = {
    async up(queryInterface) {
        const companies = await queryInterface.bulkInsert('companies', [
            { name: 'Example Company 1' },
            { name: 'Example Company 2' },
        ]);
        console.log(companies);
    },
};
```

### What do you expect to happen?

_I wanted to receive array of new records_

```
[1, 2]
```

### What is actually happening?

_The output was only single id (first in bulk)_

```
1
```

### Environment

- Sequelize version: 6.17.0
- Node.js version: 16.4.0
- Database & Version: mysql 8
- Connector library & Version: mysql2 2.3.3

## Would you be willing to resolve this issue by submitting a Pull Request?

- [ ] Yes, I have the time and I know how to start.
- [ ] Yes, I have the time but I don't know how to start, I would need guidance.
- [x] No, I don't have the time, although I believe I could do it if I had the time...
- [ ] No, I don't have the time and I wouldn't even know how to start.

**Comment by ephys** [comment]
[This PR](https://github.com/sequelize/sequelize/pull/6637) is the reason why `queryInterface.bulkInsert` is returning a single value instead of an array of values. 
I don't know the reasons behind it, they don't seem to be explained

**Comment by fzn0x** [comment]
I think it's worth to add 😃

**Comment by melnikaite** [comment]
When I use function containing `bulk` in name I expect receiving an array without any options
No any options are needed here. If I want to get first element I will use `[0]`

**Comment by WikiRik** [comment]
Yeah, I think that returning the array is the way to go. Can't really think of good reasons why you only would return a single value. Maybe if you insert a really big amount of records it might not be very feasible performance-wise to do this.

So maybe we should make the default return the array and add an optional parameter to only return the first so you know it has succeeded? Would be a breaking change, but we can still incorporate that into v7 if we get a PR for this

**Comment by ephys** [comment]
> Maybe if you insert a really big amount of records it might not be very feasible performance-wise to do this.

The array is still created so I don't think there is any difference in terms of performance 

> add an optional parameter to only return the first so you know it has succeeded

I'd prefer to avoid different return signatures based on options as much as possible, they make typings brittle

I think it's better to do the breaking change and use `results.length` or `results[0]` instead of using an option

PR welcome :)

**Comment by evanrittenhouse** [comment]
@ephys I can take this one. I know `mysql` is the marked DB type, but this should affect all of them, no?

**Comment by WikiRik** [comment]
> @ephys I can take this one. I see `mysql` is the marked DB type, but this should affect all of them, no?

Go ahead! We indeed want the same behaviour over the various dialects so this affects them all

**Comment by evanrittenhouse** [comment]
@ephys sorry, got really busy with stuff and just starting to look into this. Unless I'm wrong, this just looks like changing the return value of `bulkInsert` from `results[0]` to `results`, then update tests? Do we have a test DB to verify output against?

**Comment by ephys** [comment]
@evanrittenhouse don't worry about that, you're welcome to work on things on your own terms. We appreciate the contributions nonetheless :)

As for what needs to be done: that should the main change, but we also need to make sure that the parts of Sequelize that are using `queryInterface#bulkInsert` still work properly, as well as update the typing.

There may be tests that need to be updated, but if there aren't, could you also look into adding a simple test for that method?

**Comment by evanrittenhouse** [comment]
@ephys Cool. Read through the PR you linked [here](https://github.com/sequelize/sequelize/issues/14325#issuecomment-1089276253), and I don't necessarily see a reason for the returned array to include the number of rows affected. As far as I understand that PR, optimistic locking just adds the version number to the model - the number of rows affected as seen [here](https://github.com/sequelize/sequelize/blob/6dca5aae8b541a8c3e4028d8c4924675dc1096f3/src/dialects/sqlite/query.js#L217) may not  need to be involved, in my opinion, at least not without further context. Would you happen to be able to provide any?

**Comment by ephys** [comment]
> Would you happen to be able to provide any?

Not really, in my opinion we can do the change, and see where tests break :)

**Comment by evanrittenhouse** [comment]
@ephys I'm a little confused on the example given above. I don't see this behavior with SQLite tests, but before installing MySQL and going from there, do you know where else this issue could come up?

**Comment by WikiRik** [comment]
@evanrittenhouse bit cheaty, but I would say that you could add/update tests for this behaviour and make a draft PR so the CI can tell you for which dialects it goes wrong

**Comment by ephys** [comment]
@evanrittenhouse If you have docker installed (recommended), you can run `yarn start-mysql` to have a mysql instance ready to go :)

**Comment by evanrittenhouse** [comment]
@ephys It looks like this behavior actually differs between dialects. [Here](https://github.com/sequelize/sequelize/blob/9950b4bee896e9b50f6752be2767e28c9ba84a12/src/dialects/abstract/query.js#L83), the insert ID field is `insertId`. In SQLite, though, there's a different insert ID - `lastInsertID`. These return two different values. Using this test:
```
it.only('test', async function () => {
  await this.sequelize.queryInterface.createTable('companies', { name: DataTypes.STRING });
  const companies = await this.sequelize.queryInterface.bulkInsert('companies', [
    { name: 'Example Company 1' },
    { name: 'Example Company 2' }
  [);

  expect(companies).to.be.eq([1, 2]);
}
```
SQLite returns 2, and MySQL returns 0. Do you know the reasoning behind using the `getInsertIdField()` call to return the result? It seems like that's what's causing the single-ID return case - I'd like to preserve existing logic as much as possible, but I'm not sure to what extent it is. Would love your thoughts.

**Comment by ephys** [comment]
Checking the result of `bulkInsert`

without `returning: true`:
- postgres: `[ [], 2 ]`
- mssql: `[ [], 2 ]`
- db2: `[ [], 0 ]`
- sqlite: `[ 2, 2 ]`
- mysql: `[ 0, 2 ]`
- mariadb: `[ 0, 2 ]`

with `returning: true`
- postgres: `[ [ { name: 'Example Company 1' }, { name: 'Example Company 2' } ], 2 ]`
- mssql: `[ [ { name: 'Example Company 1' }, { name: 'Example Company 2' } ], 2 ]`
- db2: `[ [ { name: 'Example Company 1' }, { name: 'Example Company 2' } ], 2 ]`
- sqlite: `[ 2, 2 ]`
- mysql: `[ 0, 2 ]`
- mariadb: `[ 0, 2 ]`

The result in a tuple where the second entry is "how many rows were affected", and the first entry is dialect dependant.

- For postgres, mssql, & db2, it's the inserted rows (assuming `returning: true` is set)
- For sqlite, it's how many items are currently in the table
- For mysql/mariadb, it's 0 for some reason

[my interrogation from April](https://github.com/sequelize/sequelize/issues/14325#issuecomment-1089276253) is wrong: it's not returning the first inserted row and discarding the remaining rows.

It looks like the issue is that support for `RETURNING` has not been implemented in mysql/mariadb/sqlite. The reason behind it being that these databases have only added support for it "recently" (end of 2019 for mariadb).

The solution is to implement support for `RETURNING` in these dialects

**Comment by TannerS** [comment]
Is this ever implemented or planned too?

**Comment by marcosdipaolo** [comment]
+1

**Comment by notAxion** [comment]
+1

**Comment by epiplometaforiki** [comment]
After this ticket was opened (in Apr 5, 2022) it's still an issue.

**Comment by YussefSalah** [comment]
Hi, I’d like to start working on this issue. From what I understand, the change involves updating queryInterface.bulkInsert so it consistently returns an array of inserted IDs/records across dialects, along with adjusting typings and tests.

I’ll start by exploring how RETURNING is handled across different dialects and check where support might need to be added. Please let me know if there are any guidelines or preferred approaches before I dive in.

Thanks!

**Comment by LeonxLJX** [comment]
I'd like to take this one.

From the report, the useful first step is to pin down the exact code path involved before touching anything — these tend to have a small true cause behind a broad-looking symptom, and guessing at it usually produces a patch that masks the behavior instead of fixing it.

Plan: trace the path, reproduce it, then make a focused change with a test if this area has coverage. I'll keep the diff small so it's easy to review.

If there's an existing convention in this repo for how this kind of thing is handled, I'd rather follow it than invent my own — happy to be pointed at one.

(claiming via @LeonxLJX)
